File: Solutions/Remove_Duplicate_Letters/Remove_Duplicate_Letters.cs
public class Solution {
    public string RemoveDuplicateLetters(string s) {
        var mp = new Dictionary<char, int>();
        var res = new List<char>();
        
        for(int i = 0; i < s.Length; i++) {
            if(mp.Keys.Contains(s[i]))
                mp[s[i]]++;
            else
                mp.Add(s[i], 1);
        }
        
        for(int i = 0; i < s.Length; i++) {
            if(!res.Contains(s[i])) {
                while(res.Count > 0 && res[res.Count - 1] >= s[i] && mp[res[res.Count - 1]] > 0){
                    res.RemoveAt(res.Count - 1);
                }
                res.Add(s[i]);
            }
            mp[s[i]]--;
        }
        var ans = new string(res.ToArray());
        return ans;
    }
}

//Third fastest
public class Solution {
  public string RemoveDuplicateLetters (string s) {
    string unique = ""; 
    for (int i = 0; i < s.Length; i ++) {
      if (unique.Contains(s[i])) continue;

      while (unique.Length > 0 && unique[unique.Length - 1] > s[i] && s.Substring(i, s.Length - i).Contains(unique[unique.Length - 1])) { 
        unique = unique.Remove(unique.Length - 1, 1); 
        };
      unique += s[i].ToString();
    }
    // Console.Write(unique);
    return unique;
  }
}

//Second fastest
public class Solution {
    public string RemoveDuplicateLetters(string s) {
        var lastIndex = new int[26];
        for (var i = 0; i < s.Length; i++)
            lastIndex[s[i] - 'a'] = i;
        var used = new bool[26];
        for (var i = 0; i < 26; i++)
            used[i] = false;
        var res = new Stack<char>();
        for (var i = 0; i < s.Length; i++)
        {
            if(used[s[i]-'a']) continue;
            while (res.Count > 0 && res.Peek() > s[i] && lastIndex[res.Peek() - 'a'] > i)
            {
                used[res.Peek() - 'a'] = false;
                res.Pop();
            }
            used[s[i] - 'a'] = true;
            res.Push(s[i]);
        }
        var result = "";
        while (res.Count > 0)
            result = res.Pop() + result;
        return result;
    }
}

//Fastest
public class Solution {
    public string RemoveDuplicateLetters(string s) {
        var charAndCount = new int[256];

        foreach (var c in s) {
            charAndCount[c]++;
        }

        var isVisited = new HashSet<char>();
        var charStack = new Stack<char>();

        foreach (var c in s) {
            if (!isVisited.Contains(c)) {
                while (charStack.Any()) {
                    var preChar = charStack.Peek();

                    if (charAndCount[preChar] > 0 && preChar > c) {
                        var popped = charStack.Pop();
                        isVisited.Remove(popped);
                    } else {
                        break;
                    }
                }

                isVisited.Add(c);
                charStack.Push(c);
            }

            charAndCount[c]--;
        }

        var result = new List<char>();

        while (charStack.Any()) {
            result.Add(charStack.Pop());
        }

        result.Reverse();

        return string.Join("", result);
    }
}

File: Solutions/Remove_Duplicate_Letters/Remove_Duplicate_Letters.py
class Solution:
    def removeDuplicateLetters(self, s: str) -> str:
        res = []
        remaining = Counter(s)
        
        for c in s:
            if c not in res:
                while res and res[-1] >= c and remaining[res[-1]] > 0:
                    res.pop()
                res.append(c)
            remaining[c]-=1
        
        return "".join(res)

#Third fastest
class Solution:
    def removeDuplicateLetters(self, s: str) -> str:
        if not s:
            return ""
        n = len(s)
        d = collections.defaultdict(list)
        for i in range(n):
            c = s[i]
            d[c].append(i)
            
        stack = []
        for i in range(n):
            c = s[i]
            if i > d[c][0]:
                continue
            while stack and stack[-1] > c:
                t = stack[-1]
                if d[t][-1] > i:
                    stack.pop()
                    while d[t] and d[t][0] < i:
                        d[t].pop(0)
                else:
                    break

            if i == d[c][0]:
                stack.append(c)
                        
        return "".join(stack)

#Second fastest
class Solution:
    def removeDuplicateLetters(self, s: str) -> str:
        
        """
        cdadabcc
        
        01234567
        43131244
        
        1: 2, 4
        2: 5
        3: 1, 3
        4: 0, 6, 7
        
        stack = []
        c d 
        a d b c 
        
        """
        
        last_index = {}
        
        for i, char in enumerate(s):
            last_index[char] = i
            
        stack = []
        used  = set()
        
        i = 0
        while i < len(s):
            curr_char = s[i]
            if curr_char not in used:
                while stack and stack[-1] > curr_char and i < last_index[stack[-1]]:
                    used.discard(stack.pop())
                used.add(curr_char)
                stack.append(curr_char)
            i += 1
        
        return ''.join(stack)


#Fastest
class Solution:#use dict to record the last index of char,then traverse,while res里最后一个字母比char大并且后面也会出现，则pop，直至char比res的最后一个小或者最后一个是最后一次出现，确保res记录的是最每个字母必须出现一次的原则下的升序排列，则append char
    def removeDuplicateLetters(self, s: str) -> str:
        d={c:i for i,c in enumerate(s)}
        res=""
        for i,x in enumerate(s):
            if x not in res:
                while res and  x<res[-1] and i<d[res[-1]]:
                    res=res[:-1]
                res+=x
        return res